compiler/resolution.common.jvm/src/org/jetbrains/kotlin/load/java/structure/impl/classFiles/BinaryJavaClass.kt

```diff
@@ -32,13 +32,15 @@ import java.text.CharacterIterator
 import java.text.StringCharacterIterator
 
 class BinaryJavaClass(
-    override val virtualFile: VirtualFile,
+    /** If virtualFile is not available, provide classContent & innerClassFinder below. */
+    override val virtualFile: VirtualFile? = null,
     override val fqName: FqName,
     internal val context: ClassifierResolutionContext,
     private val signatureParser: BinaryClassSignatureParser,
     override var access: Int = 0,
     override val outerClass: JavaClass?,
-    classContent: ByteArray? = null
+    classContent: ByteArray? = null,
+    private val innerClassFinder: ((Name) -> JavaClass?)? = null
```

Comment:
Can you maybe use delegation? E.g
```
class BinaryJavaClassForIncremental (private val binaryJavaClass; BinaryJavaClass): JavaClass by binaryJavaClass {
       override fun findInnerClass(name: Name): JavaClass? { ... }
}
```

In such way we avoid adding third way to locate inner JavaClass to this class.

Reply:
That's a good idea. However, when I tried to do it, I ran into this issue:
  - `BinaryJavaClass.virtualFile` was originally not-null. I still have to make it nullable, because we don't have a `VirtualFile` to pass to it, only the class contents (the code to create a `VirtualFile` requires a lot of stuff that is not accessible from here).
  - Because `BinaryJavaClass.virtualFile` is nullable, in `BinaryJavaClass.findInnerClass()`, we have to handle the null case, which requires some kind of delegate, so we'll still need `innerClassFinder`. We can't ignore the null case in `BinaryJavaClass` because there's no way to enforce a subclass to override `BinaryJavaClass.findInnerClass()` to handle that case.

Reply:
kk, we can leave this as-is for now, and check with the compiler team if there are way to improve this later.